server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -421,16 +421,18 @@ private static void parseObject(final DocumentParserContext context, String curr
         Mapper objectMapper = context.getMapper(currentFieldName);
         if (objectMapper != null) {
             context.path().add(currentFieldName);
-            if (objectMapper instanceof ObjectMapper objMapper) {
-                if (objMapper.subobjects() == false) {
-                    context.path().setWithinLeafObject(true);
-                }
-            }
             parseObjectOrField(context, objectMapper);
-            context.path().setWithinLeafObject(false);
             context.path().remove();
         } else {
-            parseObjectDynamic(context, currentFieldName);
+            DynamicTemplate dynamicTemplate = context.findDynamicTemplate(currentFieldName, DynamicTemplate.XContentFieldType.OBJECT);
+            if (context.parent().subobjects() || dynamicTemplate != null) {
```

Comment:
This is to only apply dynamic templates to objects, but skip creating the  object if there is no matching template when subobjects is false? 

I think I would adapt the conditional to that, and only do something special if subobjects is false. Is it needed to look up a dynamic template externally if subobjects is true? In that case you can call parseObjectDynamic directly?

I also wonder if a cleaner solution would be to add an argument to parseObjectDynamic to control whether it should only apply the template. This will probably make it a bigger change which may warrant pulling it into a separate PR.

Reply:
hey @piergm thought about this comment?

Reply:
I do agree on just doing 

> something special if subobjects is false  
I wanted to avoid to much nesting and by adding the or clause did prevent this but on the other and as you pointed out we are looking up dynamic templates even if subobject is set to false. I'll correct this.
----
But I am not so sure that:

> a cleaner solution would be to add an argument to parseObjectDynamic to control whether it should only apply the template.

I'll try it out but does not resonate well with me.
----
I am indeed bypassing the dynamic behaviour here. Is not what we wanted in the first case? I'll add further tests to cover the case of dynamic false/runtime and subobject false.
----
> For reference: within parseObjectDynamic, we already handle dynamic:runtime, and that is very similar to what we need to do for subobjects false, as runtime mappings effectively don't have objects, and are flattened using their full path name as their field name.

Good point. I wander if there is a latent bug due to this that I solved by introducing dottedFieldNames